[2025-02 CWG Motion 9] P1967R14 `#embed`

AlisdairM opened the pull request:
Fixes https://github.com/cplusplus/draft/issues/7659.
Also fixes https://github.com/cplusplus/papers/issues/700.

Redistributed grammar around the preamble to better maintain a locality and flow.  More redistrubution expected in a follow-up commit.

Minor grammar tweak: "other than a" -> "except:"
following the model of such specifications in
basic-s-char and elsewhere.

Attempted to describe all implementation-defined
topics for the index, but will need review.

Eisenwave commented on source/preprocessor.tex:
```diff
@@ -368,6 +422,34 @@
+The parenthesized \grammarterm{pp-balanced-token-seq} in each contained
+\grammarterm{has-embed-expression} is processed as if that
+\grammarterm{pp-balanced-token-seq} were the \grammarterm{pp-tokens} in the
+third form of a \tcode{\#embed} directive\iref{cpp.embed}. If such a directive
```
Semantic line break would be nice here. I.e. prefer to put separate sentences on new lines.

Eisenwave commented on source/preprocessor.tex:
```diff
@@ -752,6 +834,356 @@
+\begin{ncsimplebnf}
+\terminal{\# embed} pp-tokens new-line
+\end{ncsimplebnf}
+(that does not match one of the two previous forms) is permitted. The
+preprocessing tokens after \tcode{embed} in the directive are processed just
+as in normal text (i.e., each identifier currently defined as a macro name is
+replaced by its replacement list of preprocessing tokens). The directive
+resulting after all replacements of the third form shall match one of the two
+previous forms.
+\begin{note}
+Adjacent \grammarterm{string-literal}{s} are not concatenated into a single
+\grammarterm{string-literal} (see the translation phases in \iref{lex.phases});
+thus, an expansion that resultsin two \grammarterm{string-literal}{s} is an
+invalid directive.
+\end{note}
+
+Any further processing as in normal text described for the two previous
+forms is not performed.
+\begin{note}
+That is, processing as in normal text happens once and only once for the entire
+directive.
+\end{note}
+
+\begin{example}
+If the directive matches the third form, the whole directive is replaced. If
+the directive matches the first two forms, everything after the name is
+replaced.
```
Semantic line breaks in this area.

Eisenwave commented on source/preprocessor.tex:
```diff
@@ -752,6 +834,356 @@
+specifies the
+maximum possible number of elements in the comma-delimited list. It shall
+appear at most once in the \grammarterm{embed-parameter-seq}. The token
+\tcode{defined} shall not appear in the constant-expression.
```
```suggestion
\tcode{defined} shall not appear in the \grammarterm{constant-expression}.
```

Eisenwave commented on source/preprocessor.tex:
```diff
@@ -752,6 +834,356 @@
+\tcode{defined} shall not appear in the constant-expression.
+
+\pnum
+The \grammarterm{pp-balanced-token-seq} is evaluated as a constant-expression
```
```suggestion
The \grammarterm{pp-balanced-token-seq} is evaluated as a \grammarterm{constant-expression}
```

AlisdairM commented on source/preprocessor.tex:
```diff
@@ -752,6 +834,356 @@
+\terminal{\# embed "} q-char-sequence \terminal{"} \opt{pp-tokens} new-line
+\end{ncsimplebnf}
+searches for a resource identified by the specified sequence between the
+\tcode{"} delimiters. The named resource is searched for in an
+\impldef{manner of search for named resource}
+manner. If this search is not supported, or if the
+search fails, the directive is reprocessed as if it read
```
This wording exactly follows the form and function of the previous subclause for `#include`.  Any suggestions for semantic line breaks should apply equally to both, and best handled by follow-up PRs after landing the motions.

jensmaurer replied:
"Semantic line break" means that a new line starts in the LaTeX source code at that point. This aids future changes, because diffs will likely be smaller.  Please read https://sembr.org/

Adding a newline there will not change the resulting rendered PDF (you'd need two newlines to make a new LaTeX paragraph, which we don't want).

Given that this is entirely LaTeX source presentation with no bearing on the PDF, any "local consistency" arguments appear not to bear much weight.  Please add the newlines.

AlisdairM replied:
So every new sentence should begin on a new line in the LaTeX source, but not in the rendered pdf?  I think I see what you mean, and can consistently make that change now.

AlisdairM commented on source/preprocessor.tex:
```diff
@@ -752,6 +834,356 @@
+
+\pnum
+\recommended The value of each integer literal should closely
+represent the bit stream of the resource unmodified. This can require an
```
I do not see how a semantic line-break helps here.  In particular, moving the `this` pronoun to the start of a new paragraph would seem to reduce clarity, or require more extensive rewording to bind clearly.

hubert-reinterpretcast commented:
Can we add _has-embed-expression_&zwnj;s to this list despite it being a change missing in the paper?
https://github.com/cplusplus/draft/blob/aece602c3ee6f4fb6adb9c0226784eb1176bc6e1/source/preprocessor.tex#L542-L547

Or do we need @jensmaurer to create a new Core Issue?

cpplearner commented on source/preprocessor.tex:
```diff
@@ -600,10 +683,8 @@
 \end{example}
 
 \rSec1[cpp.include]{Source file inclusion}
-\indextext{preprocessing directive!header inclusion}
-\indextext{preprocessing directive!source-file inclusion}
-\indextext{inclusion!source file|see{preprocessing directive, source-file inclusion}}%
-\indextext{\idxcode{\#include}}%
+\indextext{preprocessing directive!embed a resource}
+\indextext{\idxcode{\#embed}}%
```
Misplaced index?

jensmaurer replied:
Yup, that indexing change is misguided.

cpplearner commented on source/preprocessor.tex:
```diff
@@ -752,6 +833,364 @@
 \end{example}
 
+\rSec1[cpp.embed]{Resource inclusion}
+\indextext{preprocessing directive!header inclusion}
+\indextext{preprocessing directive!source-file inclusion}
+\indextext{inclusion!source file|see{preprocessing directive, source-file inclusion}}%
+\indextext{\idxcode{\#include}}%
```
Misplaced index?

jensmaurer replied:
Same here. Cut&paste going  wrong.

jensmaurer commented:
I think the change suggested by @hubert-reinterpretcast is editorial; this is part of the necessary evaluation semantics. Please add in a separate commit with the "[label] title" description style.

jensmaurer commented on source/preprocessor.tex:
```diff
@@ -45,10 +45,36 @@
     \terminal{\#} conditionally-supported-directive
 \end{bnf}
 
+\begin{bnf}
+\nontermdef{embed-parameter}\br
+    embed-standard-parameter\br
+    embed-prefixed-parameter
+\end{bnf}
+
+\begin{bnf}
+\nontermdef{embed-parameter-seq}\br
+    embed-parameter \opt{embed-parameter-seq}
+\end{bnf}
+
+\begin{bnf}
+\nontermdef{embed-standard-parameter}\br
+    \terminal{limit} \terminal{(} pp-balanced-token-seq \terminal{)}\br
+    \terminal{prefix} \terminal{(} \opt{pp-balanced-token-seq} \terminal{)}\br
+    \terminal{suffix} \terminal{(} \opt{pp-balanced-token-seq} \terminal{)}\br
+    \terminal{if_empty} \terminal{(} \opt{pp-balanced-token-seq} \terminal{)}
+\end{bnf}
+
+\begin{bnf}
+\nontermdef{embed-prefixed-parameter}\br
+    identifier :: identifier\br
+    identifier :: identifier \terminal{(} \opt{pp-balanced-token-seq} )
+\end{bnf}
+
```
Since we don't mention the (detailed) grammar elsewhere in the grammar non-terminals (we just have pp-token for #embed), please move down to immediately before or immediately after the pp-balanced-token stuff.

AlisdairM replied:
I *think* I have done as asked.

jensmaurer commented on source/preprocessor.tex:
```diff
@@ -45,10 +45,36 @@
+    embed-prefixed-parameter
+\end{bnf}
+
+\begin{bnf}
+\nontermdef{embed-parameter-seq}\br
+    embed-parameter \opt{embed-parameter-seq}
+\end{bnf}
```
This weirdly switches between forward references and no forward references.  Please move embed-parameter-seq to immediately before embed-parameter.

jensmaurer commented on source/preprocessor.tex:
```diff
@@ -752,6 +833,364 @@
+If the \gterm{implementation-resource-width} is not an integral multiple of
+\libmacro{CHAR_BIT}, the program is ill-formed.
+Let \gterm{implementation-resource-count} be
+\gterm{implementation-resource-width} divided by \libmacro{CHAR_BIT}.
```
We consider implementation-resource-width proper defined terms; please use \defn.

And remove the italics here, because that's not the defining appearance.

jensmaurer commented on source/preprocessor.tex:
```diff
@@ -752,6 +833,364 @@
+
+\pnum
+The integer literals in the comma-delimited list correspond to
+\gterm{resource-count} consecutive calls to \tcode{std::fgetc} \iref{cstdio.syn}
```
Same here: Switch to \defn; no italics here (and there should be none in the paper, but maybe we missed an occurrence during CWG review).

AlisdairM commented:
This PR is still showing as "changes requested".  I believe I have addressed all issues, could someone point out where I have missed, or remove the status?
